### ISSUE TITLE
Loss function in this repo is different from paper

### ISSUE THREAD

**Issue by zjjMaiMai** [TheSouthFrog/stylealign#8]
Hi,

In this repo, kl_loss only calculated the 2-norm between the last layers feature for dec_down and enc_down.
How does it work? Is this implementation equivalent to the formula in paper?

**Comment by TheSouthFrog** [comment]
Hi Hang,

Your problem is related to the implementation details. Actually, the KL between two Gaussian distribution with fixed covariance is equivalent to the 2-norm distance between their means.  The L2 norm implies the distributions to be with identity covariances and that's how we implement the KL. Specifically, this implement only estimates mean, which has been proved to be more stable in prior works and not to push the model too much. 

Note that we will hopefully release our PyTorch version code after CVPR DDL, where we employ the widely used 'normal' KL term. The results are losing details comparing to this one.

Best,
Shengju

**Comment by zjjMaiMai** [comment]
Hi, Shengju
Thank you for your nice response! Other question about implementation details here. 
In paper, we feed E_style with `concatenate(image, heatmap)`, but feed just image in this repo.
What is the difference between their results?

**Comment by TheSouthFrog** [comment]
We found that adding heatmap into training will make the training objective ambiguous thus being unstable. The reason I guess could be optimizing such posterior objective doesn't reflect very well at first steps. We add heatmap in the paper in order to derive the ELBO and basic VAE formulation.

**Comment by hblflybird** [comment]
> Hi Hang,
> 
> Your problem is related to the implementation details. Actually, the KL between two Gaussian distribution with fixed covariance is equivalent to the 2-norm distance between their means. The L2 norm implies the distributions to be with identity covariances and that's how we implement the KL. Specifically, this implement only estimates mean, which has been proved to be more stable in prior works and not to push the model too much.
> 
> Note that we will hopefully release our PyTorch version code after CVPR DDL, where we employ the widely used 'normal' KL term. The results are losing details comparing to this one.
> 
> Best,
> Shengju

Any plans about releasing PyTorch version code? looking forward to it...

**Comment by TheSouthFrog** [comment]
Yes, will be released soon. Hopefully in two weeks.
